File: src/lib/storage/etag-cache.ts
import { getDb } from "@/lib/storage/db";

export interface CachedEntry<T> {
  etag: string;
  body: T;
  fetchedAt: string;
}

interface Row {
  etag: string;
  body_json: string;
  fetched_at: string;
}

export async function getCached<T>(cacheKey: string): Promise<CachedEntry<T> | null> {
  const db = await getDb();
  const rows = await db.select<Row[]>(
    "SELECT etag, body_json, fetched_at FROM etag_cache WHERE cache_key = ?",
    [cacheKey],
  );
  const row = rows[0];
  if (!row) return null;
  return {
    etag: row.etag,
    body: JSON.parse(row.body_json) as T,
    fetchedAt: row.fetched_at,
  };
}

export async function setCached<T>(cacheKey: string, etag: string, body: T): Promise<void> {
  const db = await getDb();
  const fetchedAt = new Date().toISOString();
  await db.execute(
    `INSERT INTO etag_cache (cache_key, etag, body_json, fetched_at)
     VALUES (?, ?, ?, ?)
     ON CONFLICT(cache_key) DO UPDATE SET
       etag = excluded.etag,
       body_json = excluded.body_json,
       fetched_at = excluded.fetched_at`,
    [cacheKey, etag, JSON.stringify(body), fetchedAt],
  );
}

export async function clearCache(): Promise<void> {
  const db = await getDb();
  await db.execute("DELETE FROM etag_cache");
}

File: src-tauri/src/store/etag_cache.rs
//! ETag cache: `get_cached` / `set_cached` / `clear_cache` over the
//! `etag_cache` table. Port of `src/lib/storage/etag-cache.ts`.
//!
//! Unlike the JS version (which stored the parsed body), the Rust cache stores
//! the raw response text in `body_json`; callers deserialize on read.

use crate::store::db::now_iso;
use rusqlite::{Connection, OptionalExtension};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CachedEntry {
    pub etag: String,
    pub body_json: String,
    pub fetched_at: String,
}

pub fn get_cached(conn: &Connection, cache_key: &str) -> rusqlite::Result<Option<CachedEntry>> {
    conn.query_row(
        "SELECT etag, body_json, fetched_at FROM etag_cache WHERE cache_key = ?1",
        [cache_key],
        |r| {
            Ok(CachedEntry {
                etag: r.get(0)?,
                body_json: r.get(1)?,
                fetched_at: r.get(2)?,
            })
        },
    )
    .optional()
}

pub fn set_cached(
    conn: &Connection,
    cache_key: &str,
    etag: &str,
    body_json: &str,
) -> rusqlite::Result<()> {
    conn.execute(
        "INSERT INTO etag_cache (cache_key, etag, body_json, fetched_at)
         VALUES (?1, ?2, ?3, ?4)
         ON CONFLICT(cache_key) DO UPDATE SET
           etag = excluded.etag,
           body_json = excluded.body_json,
           fetched_at = excluded.fetched_at",
        (cache_key, etag, body_json, now_iso()),
    )?;
    Ok(())
}

/// Used by the frontend's "clear cache" path, wired up in Phase 2.
#[allow(dead_code)]
pub fn clear_cache(conn: &Connection) -> rusqlite::Result<()> {
    conn.execute("DELETE FROM etag_cache", [])?;
    Ok(())
}

#[cfg(test)]
#[path = "__tests__/etag_cache.rs"]
mod tests;
